### PR TITLE
Frontends for EMF metamodel plagiarism detection

### PR DESCRIPTION
**Pull request by tsaglam**
This PR introduces language modules for EMF metamodels, thus allowing to use JPlag for the plagiarism detection for metamodels. Currently, this PR contains two modules: 

- an EMF metamodel module based on a **dynamically** created token set, where tokens are extracted for every metaclass
- an EMF metamodel module based on a **handcrafted** token set and specialized token extraction rules

The dynamic frontend extends classes from the handcrafted frontends, namely the language, parser, and metamodel visitor. This PR also extends the language interface to allow for frontends where the view in the report is not the code of the language itself but rather a view on the code. Moreover, this PR adds test cases for both frontends and thus makes some additions to the test utils.

### CODE REVIEW

File: jplag.frontend.emf-metamodel/src/main/java/de/jplag/emf/util/EMFUtil.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Utility class for EMF specific functionality.
+ * @author Timur Saglam
+ */
+public final class EMFUtil {
+
+    private static final Logger logger = LoggerFactory.getLogger(EMFUtil.class.getSimpleName());
```
[the remark this answers is not in the document]
Comment (dfuchss) [reply jplag.frontend.emf-metamodel/src/main/java/de/jplag/emf/util/EMFUtil.java:26]: Logger handles the output of SimpleName

**Review by dfuchss — COMMENTED** [review]
Some comments :)

File: jplag.frontend.emf-metamodel-dynamic/src/main/java/de/jplag/emf/dynamic/Language.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    private static final int DEFAULT_MIN_TOKEN_MATCH = 10;
+
+    public Language() {
+        super(new DynamicEcoreParser());
+    }
+
+    protected Language(EcoreParser parser) {
+        super(parser);
+    }
```
[the remark this answers is not in the document]
Comment (tsaglam) [reply jplag.frontend.emf-metamodel-dynamic/src/main/java/de/jplag/emf/dynamic/Language.java:23]: Good catch, it is left over from the removed model instance prototype.

File: jplag.frontend.emf-metamodel-dynamic/pom.xml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        <dependency>
+            <groupId>org.eclipse.emf</groupId>
+            <artifactId>org.eclipse.emf.common</artifactId>
+            <version>2.25.0</version>
+        </dependency>
+        <dependency>
+            <groupId>org.eclipse.emf</groupId>
+            <artifactId>org.eclipse.emf.ecore.xmi</artifactId>
+            <version>2.16.0</version>
+        </dependency>
```
[the remark this answers is not in the document]
Comment (tsaglam) [reply jplag.frontend.emf-metamodel-dynamic/pom.xml:41]: Afaik their versions are all over the place: https://search.maven.org/search?q=g:org.eclipse.emf
So these are the newest versions for each dependency. But I think I can get rid of the XMI dependency.

File: jplag.frontend.emf-metamodel/pom.xml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        <dependency>
+            <groupId>org.eclipse.emf</groupId>
+            <artifactId>org.eclipse.emf.common</artifactId>
+            <version>2.25.0</version>
+        </dependency>
+        <dependency>
+            <groupId>org.eclipse.emf</groupId>
+            <artifactId>org.eclipse.emf.ecore.xmi</artifactId>
+            <version>2.16.0</version>
+        </dependency>
```
[the remark this answers is not in the document]
Comment (tsaglam) [reply jplag.frontend.emf-metamodel/pom.xml:37]: I can add a property to the parent pom.

File: jplag.frontend.emf-metamodel-dynamic/pom.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (dfuchss) [reply jplag.frontend.emf-metamodel-dynamic/pom.xml]: I was not sure whether all emf dependencies should be set to `2.16.0` or how compatible they are :)

[the remark this answers is not in the document]
Comment (tsaglam) [reply jplag.frontend.emf-metamodel-dynamic/pom.xml]: Actually, I am not sure either. Seems like they are I would say.

**Pull request merged**